File: python/common/union-find.py
class UnionFind:
  def __init__(self, n):
    self.parent = list(range(0, n))

  def find(self, x):
    if self.parent[x] == x:
      return x
    self.parent[x] = self.find(self.parent[x])
    return self.parent[x]

  def connect(self, x, y):
    rootX, rootY = self.find(x), self.find(y)
    self.parent[rootX] = rootY

    return rootY

class UnionFindRank:
  def __init__(self, n):
    self.parent = list(range(0, n))
    self.rank = [1] * n

  def find(self, x):
    if self.parent[x] == x:
      return x
    self.parent[x] = self.find(self.parent[x])
    return self.parent[x]

  def connect(self, x, y):
    rootX, rootY = self.find(x), self.find(y)

    if rootX == rootY:
      return rootX
    if self.rank[rootX] < self.rank[rootY]:
      self.parent[rootX] = rootY
      self.rank[rootY] += self.rank[rootX]
      return rootY
    else:
      self.parent[rootY] = rootX
      self.rank[rootX] += self.rank[rootY]
      return rootX

File: java/common/UnionFind.java
class UnionFind {

  private int[] parent;

  public UnionFind(int n) {
    parent = new int[n];
    for (int i = 0; i < n; i++) {
      parent[i] = i;
    }
  }

  int find(int x) {
    if (x == parent[x])
      return x;
    return (parent[x] = find(parent[x]));
  }

  int connect(int x, int y) {
    int x_p = find(x);
    int y_p = find(y);
    return (parent[x_p] = y_p);
  }
}
